File: ios/Tests/BareHostTests/P2PVerifyTest.swift
import BareKit
import XCTest

/// Ticket #24 / issue #41: proves the p2p native addons (rocksdb via
/// corestore, sodium via hyperdrive, udx via hyperswarm) actually run on the
/// iOS simulator, AND that a peer's drive can be opened by key, readied, and
/// read over a real hyperswarm connection (the #41 flow — the Android photo
/// app timed out on `drive.ready()`; the worklet now covers exactly that).
/// The worklet writes `p2p-verify.ok` (or `.fail`) into its storage dir, then
/// exits — the test boots it and polls for the marker.
final class P2PVerifyTest: XCTestCase {
  func testP2PStackRunsOnSimulator() throws {
    let storageDir = FileManager.default.temporaryDirectory
      .appendingPathComponent("p2p-verify-\(UUID().uuidString)", isDirectory: true)
    try FileManager.default.createDirectory(at: storageDir, withIntermediateDirectories: true)
    defer { try? FileManager.default.removeItem(at: storageDir) }

    let worklet = try Worklet(
      configuration: Worklet.Configuration(
        memoryLimit: 128 << 20,
        assets: storageDir.appendingPathComponent("asset-cache").path
      )
    )
    worklet.start(
      name: "p2p-verify",
      ofType: "bundle",
      inBundle: .main,
      arguments: [storageDir.path]
    )

    let okMarker = storageDir.appendingPathComponent("p2p-verify.ok")
    let failMarker = storageDir.appendingPathComponent("p2p-verify.fail")
    // Headroom: the worklet now runs handshake + peer-drive replication (the
    // remote ready() may take a while on a cold CI simulator).
    let deadline = Date().addingTimeInterval(240)
    while Date() < deadline {
      if FileManager.default.fileExists(atPath: okMarker.path) {
        worklet.terminate()
        return
      }
      if FileManager.default.fileExists(atPath: failMarker.path) {
        let message = try? String(contentsOf: failMarker, encoding: .utf8)
        worklet.terminate()
        return XCTFail("p2p verify failed: \(message ?? "unknown")")
      }
      Thread.sleep(forTimeInterval: 0.5)
    }
    worklet.terminate()
    XCTFail("p2p verify did not complete within 240s")
  }
}

File: examples/android-app/src/androidTest/java/to/holepunch/bare/android/P2PVerifyTest.kt
package to.holepunch.bare.android

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Assert.fail
import org.junit.Test
import org.junit.runner.RunWith
import to.holepunch.bare.kit.Worklet
import java.io.File

/**
 * Issue #41 reproduction gate on the Android emulator: boots the p2p verify
 * worklet (`core/p2p-verify.core.ts`, packed as `p2p-verify.bundle`) — the
 * same worklet the iOS `P2PVerifyTest` and the `smoke:p2p` script run — and
 * asserts it completes. The worklet covers the exact on-device flow the
 * photo app times out on: a reader opens a peer's drive by key, `ready()`s
 * it, and reads a photo over a real hyperswarm connection (rocksdb + udx +
 * sodium all native on-device). Writes `p2p-verify.ok` (or `.fail`) into its
 * storage dir and exits.
 */
@RunWith(AndroidJUnit4::class)
class P2PVerifyTest {

    @Test
    fun p2pVerifyRunsOnEmulator() {
        val context = ApplicationProvider.getApplicationContext<Context>()
        val storageDir = File(context.cacheDir, "p2p-verify-${System.currentTimeMillis()}")
        storageDir.mkdirs()

        val worklet = Worklet(
            Worklet.Options()
                .memoryLimit(128 * 1024 * 1024)
                .assets(File(storageDir, "asset-cache").absolutePath),
        )

        try {
            context.assets.open("p2p-verify.bundle").use { bundle ->
                worklet.start("/p2p-verify.bundle", bundle, arrayOf(storageDir.absolutePath))
            }

            val okMarker = File(storageDir, "p2p-verify.ok")
            val failMarker = File(storageDir, "p2p-verify.fail")
            val deadline = System.currentTimeMillis() + 240_000
            while (System.currentTimeMillis() < deadline) {
                if (okMarker.exists()) return
                if (failMarker.exists()) {
                    fail("p2p verify failed: ${failMarker.readText()}")
                }
                Thread.sleep(500)
            }
            fail("p2p verify did not complete within 240s")
        } finally {
            worklet.terminate()
        }
    }
}
